OpenAI-DotNet/Extensions/BaseEndpointExtensions.cs

```diff
@@ -0,0 +1,132 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Net.Http;
+using System.Text;
+using System.Text.Encodings.Web;
+using System.Text.Json;
+using System.Text.Json.Nodes;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OpenAI.Extensions
+{
+    internal static class BaseEndpointExtensions
+    {
+        private const string ssePattern = @"(?:(?<type>[^:\n]*):)(?<value>[^\n]*)";
+
+        private static Regex sseRegex = new(ssePattern);
+
+        private static JsonSerializerOptions sseJsonOptions = new()
+        {
+            Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping
+        };
+
+        /// <summary>
+        /// https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events
+        /// </summary>
+        public static async Task<HttpResponseMessage> StreamEventsAsync(this OpenAIBaseEndpoint baseEndpoint, string endpoint, StringContent payload, Action<HttpResponseMessage, string> eventCallback, CancellationToken cancellationToken)
+        {
+            using var request = new HttpRequestMessage(HttpMethod.Post, endpoint);
+            request.Content = payload;
+            var response = await baseEndpoint.HttpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken).ConfigureAwait(false);
+            await response.CheckResponseAsync(false, payload, null, cancellationToken).ConfigureAwait(false);
+            await using var stream = await response.Content.ReadAsStreamAsync(cancellationToken).ConfigureAwait(false);
+            using var responseStream = baseEndpoint.EnableDebug ? new MemoryStream() : null;
+            var events = new Stack<Dictionary<string, object>>();
+            using var reader = new StreamReader(stream);
+            var isEndOfStream = false;
+
+            try
+            {
+                while (await reader.ReadLineAsync() is { } streamData)
+                {
+                    if (isEndOfStream) { break; }
+                    cancellationToken.ThrowIfCancellationRequested();
+                    if (string.IsNullOrWhiteSpace(streamData)) { continue; }
+
+                    var matches = sseRegex.Matches(streamData);
+
+                    for (var i = 0; i < matches.Count; i++)
+                    {
+                        Match match = matches[i];
+                        string type;
+                        string value;
+                        string data;
+
+                        const string comment = nameof(comment);
+                        type = match.Groups[nameof(type)].Value.Trim();
+                        // If the field type is not provided, treat it as a comment
+                        type = string.IsNullOrEmpty(type) ? comment : type;
+                        value = match.Groups[nameof(value)].Value.Trim();
+                        data = match.Groups[nameof(data)].Value.Trim();
```

Comment:
.Trim() is not compliant, see the examples given by the spec itself, here: https://html.spec.whatwg.org/multipage/server-sent-events.html#event-stream-interpretation
Note specifically the example "data:  third event" which is specified to needing to contain a single leading space character after parsing. Trim() would discard this.
Additionally, the spec specifically mentions that only U+0020 space should be discarded, but .Trim() discards all whitespace characters, not just U+0020.

Reply:
Thanks @default0!